### ISSUE TITLE
geoconnect color selection

### ISSUE THREAD

**Issue by raprasad** [IQSS/dataverse#3659]
Based on usability feedback for styling a map layer:

- Remove "un-viewable" color choices
  - e.g. blue, red
- Review with Ben if we can have custom colors
  - check for map standards

\cc @mheppler @TaniaSchlatter @dlmurphy

**Comment by raprasad** [comment]
We can customize the color ranges from the Dataverse/Geoconnect side.  
Task: For each color choice
  - create color name that user will see
  - start color range in hex format
  - end color range in hex format

**Comment by mheppler** [comment]
I did a quick search for some map color theme examples, and found a [nice resource](https://tilemill-project.github.io/tilemill/docs/guides/tips-for-color/) to start with.

![screen shot 2017-03-30 at 11 22 37 am](https://cloud.githubusercontent.com/assets/687227/24511979/47e597f2-153b-11e7-8eac-454bf64d9a82.png)

**Comment by pdurbin** [comment]
@mheppler thanks, that reminds me of the "categorical" colors D3 uses: https://github.com/d3/d3-scale-chromatic#categorical

![d3_d3-scale-chromatic_sequential _diverging_and_categorical_color_scales_derived_from_colorbrewer _-_2017-03-30_11 27 50](https://cloud.githubusercontent.com/assets/21006/24512280/2147422a-153c-11e7-8d07-073a3bf49a46.png)

**Comment by pdurbin** [comment]
Before I mess with the code too much I'm capturing below screenshots of "before" as of https://github.com/IQSS/geoconnect/commit/0aef3c5 

![screen shot 2017-03-30 at 1 14 10 pm](https://cloud.githubusercontent.com/assets/21006/24517080/e5fea1b8-154a-11e7-8979-74677b0cd30f.png)


![screen shot 2017-03-30 at 1 14 27 pm](https://cloud.githubusercontent.com/assets/21006/24517073/dedd6720-154a-11e7-9d51-2fed4f18f381.png)

**Comment by raprasad** [comment]
@pdurbin : The color choices may be changed via the admin panel (values are in database)

On your local geoconnect, go to:  http://127.0.0.1:8070/geo-connect-admin/layer_classification/colorramp/

![select_color_ramp_to_change___django_site_admin](https://cloud.githubusercontent.com/assets/1036060/24517296/b7dfed04-154b-11e7-9a4e-f6e3a06b6019.png)

**Comment by raprasad** [comment]
When you're ready to save new colors:

- view the db values:
```
python manage.py dumpdata layer_classification --indent=4
```

- save them to the fixture used for loading the colors
  - May need some discussion.  Currently stored in [shared_dataverse_information](https://github.com/IQSS/shared-dataverse-information) which should perhaps change.

      ```
      python manage.py dumpdata layer_classification --indent=4 > {file name}
      ```
  - This line loads the initial data during install:
      ```
      python manage.py loaddata --app layer_classification initial_data.json
      ```

**Comment by pdurbin** [comment]
@raprasad thanks. Here's a dump of just the current values in question ("colorramp"), thanks to `jq`:

`$ python manage.py dumpdata layer_classification --indent=4 | jq '.[] | select(.model=="layer_classification.colorramp")'`

```
{
  "model": "layer_classification.colorramp",
  "pk": 1,
  "fields": {
    "display_name": "Blue",
    "value_name": "Blue",
    "sort_order": 10,
    "start_color": "#f7fbff",
    "end_color": "#08306b",
    "active": true,
    "created": "2014-07-18T17:03:05.792Z",
    "modified": "2014-07-18T17:03:05.792Z"
  }
}
{
  "model": "layer_classification.colorramp",
  "pk": 2,
  "fields": {
    "display_name": "Red",
    "value_name": "Red",
    "sort_order": 20,
    "start_color": "#fff5f0",
    "end_color": "#67000d",
    "active": true,
    "created": "2014-07-18T17:03:05.795Z",
    "modified": "2014-07-18T17:03:05.795Z"
  }
}
{
  "model": "layer_classification.colorramp",
  "pk": 3,
  "fields": {
    "display_name": "Orange",
    "value_name": "Custom",
    "sort_order": 30,
    "start_color": "#fff5eb",
    "end_color": "#f16913",
    "active": true,
    "created": "2014-07-18T17:03:05.796Z",
    "modified": "2014-07-18T17:03:05.797Z"
  }
}
{
  "model": "layer_classification.colorramp",
  "pk": 4,
  "fields": {
    "display_name": "Red-Blue",
    "value_name": "Jet",
    "sort_order": 40,
    "start_color": "#f7fbff",
    "end_color": "#08306b",
    "active": true,
    "created": "2014-07-18T17:03:05.797Z",
    "modified": "2014-07-18T17:03:05.797Z"
  }
}
{
  "model": "layer_classification.colorramp",
  "pk": 5,
  "fields": {
    "display_name": "Gray",
    "value_name": "Gray",
    "sort_order": 50,
    "start_color": "#fff5eb",
    "end_color": "#f16913",
    "active": true,
    "created": "2014-07-18T17:03:05.798Z",
    "modified": "2014-07-18T17:03:05.798Z"
  }
}
{
  "model": "layer_classification.colorramp",
  "pk": 6,
  "fields": {
    "display_name": "Random",
    "value_name": "Random",
    "sort_order": 60,
    "start_color": "#f7fbff",
    "end_color": "#08306b",
    "active": true,
    "created": "2014-07-18T17:03:05.799Z",
    "modified": "2014-07-18T17:03:05.799Z"
  }
}
{
  "model": "layer_classification.colorramp",
  "pk": 7,
  "fields": {
    "display_name": "Custom",
    "value_name": "Custom",
    "sort_order": 70,
    "start_color": "#ffcc00",
    "end_color": "#006699",
    "active": false,
    "created": "2014-07-18T17:03:05.800Z",
    "modified": "2014-07-18T17:03:05.800Z"
  }
}
```

**Comment by pdurbin** [comment]
And I see what you mean, I'm in the wrong repo. I should be making a pull request against this file: https://github.com/IQSS/shared-dataverse-information/blob/master/shared_dataverse_information/layer_classification/fixtures/initial_data.json

I guess this means making a release to pypi. 0.5.5 is the current version of https://pypi.python.org/pypi/shared_dataverse_information

**Comment by raprasad** [comment]
For the jq line above, you can also specify specific models:
```
# colorramp
python manage.py dumpdata layer_classification.colorramp --indent=4

# classificationmethod
python manage.py dumpdata layer_classification.classificationmethod --indent=4
```





To keep the install consistent, we should keep the final 'dumpdata' to include ```colorramp``` and ```classificationmethod```

**Comment by raprasad** [comment]
Let's talk tomorrow.  e.g. Probably easiest to experiment/set colors first and the start with the other repository...

**Comment by pdurbin** [comment]
@raprasad is seeing the same trouble I am. We're stuck with dark blue. We were able to prove that the page ( http://127.0.0.1:8070/shapefile/view-shp-examine/ ) *is* paying attention to the database by toggling the "active" boolean. I also just tried adding a brand new row called "NewBlue" (rather than existing the existing "Blue") using the JSON below, but I got this error:

"Sorry! The classification failed.

(The following errors were encounted: ramp: Select a valid choice. NewBlue is not one of the available choices.)"

This error comes from here: https://github.com/IQSS/geoconnect/blob/9f3e4aa8a7e3a0c07784de4d9b03f9511ca000c3/gc_apps/classification/views.py#L192

In short, @raprasad is going to do some investigation.

![screen shot 2017-03-31 at 9 06 55 am](https://cloud.githubusercontent.com/assets/21006/24551738/a1f282a8-15f1-11e7-919a-78249846a0bf.png)


```
{
    "pk": 8, 
    "model": "layer_classification.colorramp", 
    "fields": {
        "value_name": "NewBlue", 
        "display_name": "NewBlue", 
        "created": "2014-07-18T13:03:05.800", 
        "modified": "2014-07-18T13:03:05.800", 
        "active": true, 
        "start_color": "#ffffff", 
        "end_color": "#006699", 
        "sort_order": 70
    }
}
```

**Comment by pdurbin** [comment]
I found the following list of usual suspects at https://github.com/cga-harvard/worldmap/blob/3e92c4867a82a3fc8df35d1766094c3dd689adbb/src/geonode-client/app/static/externals/gxp/src/script/widgets/ClassificationPanel.js#L18

    rampBlueText: "Blue",
    rampRedText: "Red",
    rampOrangeText: "Orange",
    rampJetText: "Blue-Red",
    rampGrayText: "Gray",
    rampRandomText: "Random",
    rampCustomText: "Custom",

**Comment by TaniaSchlatter** [comment]
Let's use this map color picker as a resource for map colors. It is based on extensive research and is the best I've found:
• http://colorbrewer2.org
• GitHub: https://github.com/axismaps/colorbrewer/

**Comment by raprasad** [comment]
Note:
- For validation, WorldMap* checks that the color ramp names sent by geoconnect are the same as names in the WorldMap tables*
- If geoserver doesn't care about color names, then:
    - No actual need for name validation.
    - Tweak WorldMap code to skip name check but continue to check for valid hex colors 

* both added by dv team and not used in other places by WorldMap

**Comment by pdurbin** [comment]
In short, this issue is blocked on production database changes being made in WorldMap and @raprasad mentioned that the WorldMap developer we'd need to talk to is on vacation all next week. I'm going to put this issue into the Backlog at https://waffle.io/IQSS/dataverse and unassign myself. There's not much I can do. If @mheppler and @TaniaSchlatter want to continue to work out which colors to use, that's fine, of course.

**Comment by raprasad** [comment]
- Codebase changes needed
  - WorldMap: Disable matching of color names and use the colors sent via url
  - Geoconnect: Update the color ramps

**Comment by raprasad** [comment]
Map color params for SLD service: http://docs.geoserver.org/stable/en/user/community/sldservice/index.html#classify-vector-data

**Comment by raprasad** [comment]
note for self: on local worldmap, check how specifying color midpoints works

**Comment by raprasad** [comment]
RP checklist
- [ ] Test color midpoints, send screenshots back

## If color midpoints work
- [ ] Update color ramp model geoconnect to include midpoints
- [ ] Update API endpoint on WorldMap to accept optional midpoints

## Cleanup (also part of/simplifies #3640)

Clean-up steps after finding that the colors are not a custom part of geonode but part of core geoserver and that the attributes for colors/algorithms are fixed/limited and shouldn't be part of db entries on the WorldMap side

- [ ] Move module from shared_dataverse_information directly to geoconnect
  - [ ] Literally copy module to geoconnect (unchanged)
  - [ ] Remove module from shared_dataverse_information
  - [ ] Update version of shared_dataverse_information and send to pypi
  - [ ] Geoconnect: update shared_dataverse_information version in the requirements file
- [ ] WorldMap:
   - [ ]  Remove classify models from settings (2 lines) 
   - [ ] Delete color ramp/classify algorithm tables
     - e.g. WorldMap doesn't need to have this and it is extra overhead
   - [ ] Updated version of shared_dataverse_information

**Comment by raprasad** [comment]
Worked with @mheppler through several color schemes including ```midColor``` (e.g. manually changing WorldMap code)
  - ```midColor``` is not having any effect
  - Will go with new/updated start/end color schemes

**Comment by raprasad** [comment]
- New color file from @mheppler checked in and applied to dev geoconnect.
- Colors available via dvn-build, e.g. https://dvn-build.hmdc.harvard.edu
/file.xhtml?fileId=2889430&version=2.0

**Comment by raprasad** [comment]
connects to: https://github.com/IQSS/geoconnect/pull/131

**Comment by pdurbin** [comment]
Pull request https://github.com/IQSS/geoconnect/pull/131 looks good to me. Moving to QA at https://waffle.io/IQSS/dataverse

**Comment by mheppler** [comment]
Discussed with @raprasad. Also changing the stroke color and width in the map. The ColorBrewer map uses `stroke-width: .50px;` and defaults to black border color. These settings will be applied to WorldMap along with the color scheme improvements outlined here.

**Comment by kcondon** [comment]
OK so actual changes are:
1. Change only current red and blue colors to have a better range/separation, done on GeoConnect but default file is checked in to Github
2. Add back black border, requires code change to World Map

**Comment by raprasad** [comment]
Worldmap PR, connects to: https://github.com/cga-harvard/cga-worldmap/pull/231

**Comment by raprasad** [comment]
Check list:
 - [ ] WorldMap PR to go live
 - [ ] Update prod. geoconnect

**Comment by pdurbin** [comment]
@raprasad and I just did code review of https://github.com/cga-harvard/cga-worldmap/pull/231 and it looks good. I'd suggest writing up the changes for @kcondon 's benefit while testing and/or describing them to him face to face. We looked at the red and blue maps and they are way more usable than the ones I took screenshots of at https://github.com/IQSS/dataverse/issues/3659#issuecomment-290479182

Future developers should note that colors do not come from https://github.com/IQSS/shared-dataverse-information/blob/master/shared_dataverse_information/layer_classification/fixtures/initial_data.json anymore. Then now come from the geoconnect repo (initial_data_2017_0421.json) and the README has already been updated: https://github.com/IQSS/geoconnect/commit/164ed8c74658dba5a289a6c31e4073e9027f3ff2 . Basically, it'll be easier to tweak these colors in the future and we won't be dependent on any changes to the WorldMap side (once the pull request above is merged).

**Comment by raprasad** [comment]
Moved to "Done".  Last change was stroke color/size for polygons made on the WorldMap side.  Just checked it with @mheppler
